File: src/main/java/org/jetbrains/fortran/lang/parser/KeywordParser.java
package org.jetbrains.fortran.lang.parser;

import com.intellij.lang.PsiBuilder;
import com.intellij.lang.parser.GeneratedParserUtilBase;

import static com.intellij.lang.parser.GeneratedParserUtilBase.*;
import static org.jetbrains.fortran.lang.FortranTypes.*;
import static org.jetbrains.fortran.lang.psi.FortranTokenType.KEYWORD;
import static org.jetbrains.fortran.lang.psi.FortranTokenType.WORD;

public class KeywordParser implements GeneratedParserUtilBase.Parser {

    private final String keyword_text;
    public KeywordParser(String keywordType) {
        keyword_text = keywordType;
    }

    @Override
    public boolean parse(final PsiBuilder builder, final int level) {
        if (!recursion_guard_(builder, level, "Identifier")) return false;
        boolean result = false;
        PsiBuilder.Marker marker = enter_section_(builder);
        if (builder.getTokenType() == IDENTIFIER || builder.getTokenType() == WORD
                || builder.getTokenType() == KEYWORD) {
            if (keyword_text.equalsIgnoreCase(builder.getTokenText())) {
                builder.remapCurrentToken(KEYWORD);
                result = consumeToken(builder, KEYWORD);
            }
        }
        exit_section_(builder, marker, null, result);
        return result;
    }
}

File: src/main/kotlin/org/jetbrains/fortran/lang/parser/KeywordParser.kt
package org.jetbrains.fortran.lang.parser

import com.intellij.lang.PsiBuilder
import org.jetbrains.fortran.lang.FortranTypes.IDENTIFIER
import org.jetbrains.fortran.lang.parser.FortranParserUtil.consumeToken
import org.jetbrains.fortran.lang.parser.FortranParserUtil.enter_section_
import org.jetbrains.fortran.lang.parser.FortranParserUtil.exit_section_
import org.jetbrains.fortran.lang.parser.FortranParserUtil.recursion_guard_
import org.jetbrains.fortran.lang.psi.FortranTokenType

class KeywordParser(private val keyword_text: String) : FortranParserUtil.Parser {
    override fun parse(builder: PsiBuilder, level: Int): Boolean {
        if (!recursion_guard_(builder, level, "Identifier")) return false
        val expectedType = FortranTokenType.getKeyword(keyword_text.toLowerCase()) ?: FortranTokenType.KEYWORD
        var result = false
        val marker = enter_section_(builder)
        val tokenType = builder.tokenType
        if (tokenType === IDENTIFIER || tokenType === FortranTokenType.WORD || FortranTokenType.KEYWORDS.contains(tokenType)) {
            if (keyword_text.equals(builder.tokenText!!, ignoreCase = true)) {
                builder.remapCurrentToken(FortranParserUtil.cloneTTwithBase(tokenType, expectedType))
                result = consumeToken(builder, expectedType)
            }
        }
        exit_section_(builder, marker, null, result)
        return result
    }
}


